src/main/osd/osd_warnings.c

```diff
@@ -277,32 +330,29 @@ void renderOsdWarning(char *warningText, bool *blinking, uint8_t *displayAttr)
 #ifdef USE_ESC_SENSOR
     // Show warning if we lose motor output, the ESC is overheating or excessive current draw
     if (featureIsEnabled(FEATURE_ESC_SENSOR) && osdWarnGetState(OSD_WARNING_ESC_FAIL) && ARMING_FLAG(ARMED)) {
-        char* p = warningText;
-        strcpy(p, "ESC");
-        p += strlen("ESC");
+        uint8_t escErrorLength = 0;
+        
+        // Write 'ESC' prefix
+        escErrorLength += tfp_sprintf(warningText + escErrorLength, "ESC");
 
         bool escWarning = false;
-        for (unsigned i = 0; i < getMotorCount() && p < warningText + OSD_WARNINGS_MAX_SIZE - 1; i++) {
+        for (unsigned i = 0; i < getMotorCount() && escErrorLength < OSD_WARNINGS_MAX_SIZE - 6; i++) {
             escSensorData_t *escData = getEscSensorData(i);
-            // if everything is OK just display motor number else R, T or C
-            if (osdConfig()->esc_current_alarm != ESC_CURRENT_ALARM_OFF
-                && escData->current >= osdConfig()->esc_current_alarm) {
-                *p++ = 'C';
-                escWarning = true;
-            } else if (osdConfig()->esc_temp_alarm != ESC_TEMP_ALARM_OFF
-                       && escData->temperature >= osdConfig()->esc_temp_alarm) {
-                *p++ = 'T';
+
+            char alarmChars[4]; // Buffer for alarm characters (C, T, R) OR motor number (up to 2 digits)
+
+            if (checkEscAlarmConditions(i, erpmToRpm(escData->rpm), escData->temperature, escData->current, alarmChars)) {
```

Comment:
This should append both with and without alarm - either error or motor number. Only remember escWarning to drop all collected numbers when there is no alarm

Reply:
Also note in original ESC_SENSOR code we only add one alarm char as with dshot we use all.

https://github.com/betaflight/betaflight/blob/master/src/main/osd/osd_warnings.c#L287-L302

Reply:
IMO is is mostly poor reimplementation of identical feature - is there any reason to make difference between ESC_SENSOR and dshot (or any other per-motor telemetry)?

Reply:
Agree, unified the two in a single function now.